File: Cluster/egglib-3.0.0b21/pyfiles/stats/_filter.py
__license__ = """
    Copyright 2015-2016 Stephane De Mita, Mathieu Siol

    This file is part of EggLib.

    EggLib is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    EggLib is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with EggLib.  If not, see <http://www.gnu.org/licenses/>.
"""

from .. import _eggwrapper

########################################################################

class Filter(object):

    """
    Holds lists of valid (exploitable and missing) data codes. The
    constructor allows to immediately instanciate the class with all
    desired information. It is also possible to call the modifier
    methods to add information. Note that this module provides a few
    pre-defined instances with common settings that the user may use and
    possibly extend.

    .. note::
        If no codes are specified (which is the default behaviour if no
        arguments are specified), the filter will treat all codes as
        acceptable.

    :param exploitable: Value or values used as exploitable data.
        Possible values are ``None``, a single integer, a list (or other
        iterable) of integers, or a string. An iterable of one-character
        strings is also possible. Negative integers are supported.

    :param rng: A range of integer values taken as exploitable. Must
        contain two values, providing the minimum and maximum values.

    :param missing: Value or values used as missing data. Specifications
        as for *exploitable*.

    :param lower_alias: For all *exploitable* and *missing* (but not
        *rng*) values, use the lower case character as alias. Ignored
        for all integer values.

    :param exploitable_alias: Provide the alias values for exploitable
        data. The number of values and their order must match the
        argument *exploitable*.

    :param missing_alias: Provide the alias values for missing data. The
        number of values and their order must match the argument
        *missing*.
    """

    ####################################################################

    @staticmethod
    def _process(array):
        if isinstance(array, int): return [array]
        try: return [i if isinstance(i, int) else ord(i) for i in array]
        except TypeError, e: raise ValueError, 'invalid data: invalid type for exploitable/missing value found: {0}'.format(e.message)

    ####################################################################

    def __init__(self, exploitable=None, rng=None, missing=None,
                 exploitable_alias=None, missing_alias=None,
                 lower_alias=False):

        self._obj = _eggwrapper.Filter()

        if exploitable != None:
            self.exploitable(exploitable, alias=exploitable_alias, lower_alias=lower_alias)
        elif exploitable_alias != None:
            raise ValueError, 'cannot specify `exploitable_alias` if `exploitable` is `None`'

        if missing != None:
            self.missing(missing, alias=missing_alias, lower_alias=lower_alias)
        elif missing_alias != None:
            raise ValueError, 'cannot specify `missing_alias` if `missing` is `None`'

        if rng != None:
            try:
                if len(rng) != 2: raise ValueError, 'invalid `rng` argument: must have two items'
            except TypeError: raise TypeError, 'invalid `rng` argument'
            self.rng(*rng)

    ####################################################################

    def exploitable(self, values, alias=None, lower_alias=False):  # almost identical copies . . .

        """
        Add exploitable values. See the class documentation for more
        details.
        """


        if lower_alias == True and alias != None:
            raise ValueError, 'cannot specify both `alias` and `lower_alias` at the same time'

        values = self._process(values)
        if alias == None:
            for a in values:
                if lower_alias == True and a>=65 and a<=90:
                    self._obj.add_exploitable_with_alias(a, a+32)
                else:
                    self._obj.add_exploitable(a)
        else:
            alias = self._process(alias)
            if len(values) != len(alias): raise ValueError, 'length of `values` and `alias` don\'t match'
            for a, b in zip(values, alias): self._obj.add_exploitable_with_alias(a, b)

    ####################################################################

    def missing(self, values, alias=None, lower_alias=False): # almost identical copies . . .

        """
        Add missing values. See the class documentation for more
        details.
        """

        if lower_alias == True and alias != None:
            raise ValueError, 'cannot specify both `alias` and `lower_alias` at the same time'

        values = self._process(values)
        if alias == None:
            for a in values:
                if lower_alias == True and a>=65 and a<=90:
                    self._obj.add_missing_with_alias(a, a+32)
                else:
                    self._obj.add_missing(a)
        else:
            alias = self._process(alias)
            if len(values) != len(alias): raise ValueError, 'length of `values` and `alias` don\'t match'
            for a,b in zip(values, alias): self._obj.add_missing_with_alias(a, b)

    ####################################################################

    def rng(self, mini, maxi):

        """
        Add a range of exploitable data. See the class documentation for
        more details.
        """

        if mini > maxi: raise ValueError, 'minimum value of a range of exploitable values must not be greater than maximum'
        mini, maxi = self._process([mini, maxi])
        self._obj.add_exploitable_range(mini, maxi)

########################################################################

#: All possible allele values are accepted as valid.
#: This filter is case-sensitive.
filter_default = Filter()

#: Used for DNA sequences (case-independent).
#: Exploitable: ``ACGT``. Missing: ``RYSWKMBDHVN?-``.
#: Lower-case characters are mapped to the corresponding upper-case character.
filter_dna = Filter(exploitable='ACGT', missing='RYSWKMBDHVN?-', lower_alias=True)

#: Used for RNA sequences (case-independent).
#: Exploitable: ``ACGU``. Missing: ``RYSWKMBDHVN?-``. The IUPAC codes for
#: missing data are identical in meaning to those for DNA sequences.
#: Lower-case characters are mapped to the corresponding upper-case character.
filter_rna = Filter(exploitable='ACGU', missing='RYSWKMBDHVN?-', lower_alias=True)

#: Paranoid filter for DNA sequences. Exploitable: ``ACGT``. Missing: ``N``.
#: Lower-case characters are not allowed.
filter_strict = Filter(exploitable='ACGT', missing='N', lower_alias=False)

#: Used for strictly positive SSR alleles.
#: Exploitable: range 1--999. Missing: -1.
filter_ssr = Filter(rng=(1, 999), missing=-1)

#: Used for numerical values.
#: Exploitable: range -999--999. Missing: none.
filter_num = Filter(rng=(-999, 999))

#: Used for amino acid sequences. Exploitable: ``ACDEFGHIKLMNPQRSTVWY``.
#: Missing: ``X-``. 
#: Lower-case characters are mapped to the corresponding upper-case character.
filter_amino = Filter(exploitable='ACDEFGHIKLMNPQRSTVWY', missing='X-', lower_alias=True)

#: Used for numerical codon representation. Exploitable: range 0--63. Missing: 64.
filter_codon = Filter(rng=(0, 63), missing=64)

File: Cluster/egglib-3.0.0b21/cppfiles/Filter.cpp
/*
    Copyright 2012-2013 Stéphane De Mita, Mathieu Siol
    
    This file is part of the EggLib library.

    EggLib is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    EggLib is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with EggLib.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "egglib.hpp"
#include "Filter.hpp"
#include <cstdlib>

namespace egglib {

    Filter::Filter() {
        init();
    }

    //////

    Filter::Filter(const Filter& src) {
        init();
        copy(src);
    }

    //////

    Filter& Filter::operator=(const Filter& src) {
        copy(src);
        return *this;
   }

    //////

    Filter::~Filter() {
        free();
    }

    //////

    void Filter::reserve(unsigned int num_expl, unsigned int num_expl_rg, unsigned int num_missing, unsigned int num_synonyms, unsigned int num_missing_synonyms) {

        if (num_expl > _ne_c) {
            _exploitable = (int *) realloc(_exploitable, num_expl * sizeof(int));
            if (!_exploitable) throw EGGMEM;
            _ne_c = num_expl;
        }

        if (num_expl_rg > _ne_c_rg) {
            _exploitable_rg = (int *) realloc(_exploitable_rg, 2 * num_expl_rg * sizeof(int));
            if (!_exploitable_rg) throw EGGMEM;
            _ne_c_rg = num_expl_rg;
        }

        if (num_missing > _nm_c) {
            _missing = (int *) realloc(_missing, num_missing * sizeof(int));
            if (!_missing) throw EGGMEM;
            _nm_c = num_missing;
        }
        
        if (num_synonyms > _nsyn_c) {
            _syn = (int *) realloc(_syn, 2 * num_synonyms * sizeof(int));
            if (!_syn) throw EGGMEM;
            _nsyn_c = num_synonyms;
        }

        if (num_missing_synonyms > _n_missing_syn_c) {
            _missing_syn = (int *) realloc(_missing_syn, 2 * num_missing_synonyms * sizeof(int));
            if (!_missing_syn) throw EGGMEM;
            _n_missing_syn_c = num_missing_synonyms;
        }
    }

    //////

    void Filter::clear() {
        free();
        init();
    }

    //////

    void Filter::add_exploitable(int code) {
        _ne++;
        if (_ne > _ne_c) {
            _exploitable = (int *) realloc(_exploitable, _ne * sizeof(int));
            if (!_exploitable) throw EGGMEM;
            _ne_c = _ne;
        }
        _exploitable[_ne-1] = code;
    }

    //////

    void Filter::add_exploitable_with_alias(int code, int alias) {
        add_exploitable(code);
        _nsyn++;
        if (_nsyn > _nsyn_c) {
            _syn = (int *) realloc(_syn, 2 * _nsyn * sizeof(int));
            if (!_syn) throw EGGMEM;
            _nsyn_c = _nsyn;
        }
        _syn[2 * (_nsyn-1)] = alias;
        _syn[2 * (_nsyn-1) + 1] = code;
    }
    
    //////

    void Filter::add_exploitable_range(int first, int last) {
        _ne_rg++;
        if (_ne_rg > _ne_c_rg) {
            _exploitable_rg = (int *) realloc(_exploitable_rg, 2 * _ne_rg * sizeof(int));
            if (!_exploitable_rg) throw EGGMEM;
            _ne_c_rg = _ne_rg;
        }
        _exploitable_rg[2*(_ne_rg-1)] = first;
        _exploitable_rg[2*(_ne_rg-1) + 1] = last;
    }

    //////

    void Filter::add_missing(int code) {
        _nm++;
        if (_nm > _nm_c) {
            _missing = (int*) realloc(_missing, _nm*sizeof(int));
            if (!_missing) throw EGGMEM;
            _nm_c = _nm;
        }
        _missing[_nm-1] = code;
    }

    //////

    void Filter::add_missing_with_alias(int code, int alias) {
        add_missing(code);
        _n_missing_syn++;
        if (_n_missing_syn > _n_missing_syn_c) {
            _missing_syn = (int *) realloc(_missing_syn, 2 * _n_missing_syn * sizeof(int));
            if (!_missing_syn) throw EGGMEM;
            _n_missing_syn_c = _n_missing_syn;
        }
        _missing_syn[2 * (_n_missing_syn-1)] = alias;
        _missing_syn[2 * (_n_missing_syn-1) + 1] = code;
    }
    
    //////

    bool Filter::is_exploitable(int& code) const {

        // liberal case
        if (_ne+_ne_rg==0) return true;
        
        // check if match in lonely codes
        for (unsigned int i=0; i<_ne; i++) {
            if (code == _exploitable[i]) return true;
        }

        // check if match in code ranges
        for (unsigned int i=0; i<_ne_rg; i++) {
            if (code >= _exploitable_rg[2*i] && code <= _exploitable_rg[2*i+1]) return true;
        }
        
        // check if match in synonyms
        for (unsigned int i=0; i<_nsyn; i++) {
            if (code == _syn[2*i]) {
                code = _syn[2*i + 1];
                return true;
            }
        }

        // no match
        return false;
    }

    //////

    bool Filter::is_missing(int& code) const {
        
        // check if match
        for (unsigned int i=0; i<_nm; i++) {
            if (code == _missing[i]) return true;
        }
        
        // check if match in synonyms
        for (unsigned int i=0; i<_n_missing_syn; i++) {
            if (code == _missing_syn[2*i]) {
                code = _missing_syn[2*i + 1];
                return true;
            }
        }
        
        // no match
        return false;
    }

    //////

    int Filter::check(int code, bool& flag) const {
        if (_ne+_ne_rg==0) return code;
        for (unsigned int i=0; i<_ne; i++) if (code == _exploitable[i]) return code;
        for (unsigned int i=0; i<_ne_rg; i++) if (code >= _exploitable_rg[2*i] && code <= _exploitable_rg[2*i+1]) return code;
        for (unsigned int i=0; i<_nsyn; i++) if (code == _syn[2*i]) return _syn[2*i + 1];
        for (unsigned int i=0; i<_nm; i++) if (code == _missing[i]) return MISSINGDATA;
        for (unsigned int i=0; i<_n_missing_syn; i++) if (code == _missing_syn[2*i]) return MISSINGDATA;
        flag = true;
        return MISSINGDATA;
    }

    //////

    void Filter::init() {
        _exploitable = NULL;
        _ne = 0;
        _ne_c = 0;
        _exploitable_rg = NULL;
        _ne_rg = 0;
        _ne_c_rg = 0;
        _missing = NULL;
        _nm = 0;
        _nm_c = 0; 
        _syn = NULL;
        _nsyn = 0;
        _nsyn_c = 0;
        _missing_syn = NULL;
        _n_missing_syn = 0;
        _n_missing_syn_c = 0;
    }

    //////

    void Filter::copy(const Filter& src) {

        reserve(src._ne, src._ne_rg, src._nm, src._nsyn, src._n_missing_syn);
        _ne = src._ne;
        _ne_rg = src._ne_rg;
        _nm = src._nm;
        _nsyn = src._nsyn;
        _n_missing_syn = src._n_missing_syn;

        for (unsigned int i=0; i<src._ne; i++) {
            _exploitable[i] = src._exploitable[i];
        }
        
        for (unsigned int i=0; i<src._ne_rg; i++) {
            _exploitable_rg[2*i] = src._exploitable_rg[2*i];
            _exploitable_rg[2*i+1] = src._exploitable_rg[2*i+1];
        }

        for (unsigned int i=0; i<src._nm; i++) {
            _missing[i] = src._missing[i];
        }
        
        for (unsigned int i=0; i<src._nsyn; i++) {
            _syn[2*i] = src._syn[2*i];
            _syn[2*i+1] = src._syn[2*i+1];
        }

        for (unsigned int i=0; i<src._n_missing_syn; i++) {
            _missing_syn[2*i] = src._missing_syn[2*i];
            _missing_syn[2*i+1] = src._missing_syn[2*i+1];
        }
    }
            
    //////

    void Filter::free() {
        if (_exploitable) ::free(_exploitable);
        if (_exploitable_rg) ::free(_exploitable_rg);
        if (_missing) ::free(_missing);
        if (_syn) ::free(_syn);
        if (_missing_syn) ::free(_missing_syn);
    }
}
